Classes/Domain/Variants/IdBuilder.php

```
@@ -0,0 +1,95 @@
+
+/**
+ * The variantId can be used to group documents by a variantId. This variantId is by default unique per system,
+ * and has the following syntax:
+ *
+ * <SystemHash>/type/uid
+ *
+ * A file from one system will get the same variantId, which could be useful for de-duplication.
+ * @author Timo Hund <timo.hund@dkd.de>
+ */
+class IdBuilder
+{
+
+    /**
+     * This method is used to build a variantId.
+     *
+     * By default the variantId is used
+     * @param string $type
+     * @param integer $uid
+     * @return string
+     */
+    public function buildFromTypeAndUid($type, $uid)
+    {
+        $systemHash = $this->getSystemHash();
+        $variantId = $systemHash . '/' . $type . '/' . $uid;
+
+        $variantId = $this->applyHook($variantId, $systemHash, $type, $uid);
+        return $variantId;
+    }
+
+    /**
+     * Applies configured postProcessing hooks to build a custom variantId.
+     *
+     * @param string $variantId
+     * @param string $systemHash
+     * @param string $type
+     * @param integer $uid
+     * @return string
+     */
+    protected function applyHook($variantId, $systemHash, $type, $uid)
+    {
+        if (!is_array($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['solr']['modifyVariantId'])) {
+            return $variantId;
+        }
+
+        foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['solr']['modifyVariantId'] as $classReference) {
+            $variantIdModifier = GeneralUtility::getUserObj($classReference);
+            if ($variantIdModifier instanceof IdModifier) {
+                $variantId = $variantIdModifier->modifyVariantId($variantId, $systemHash, $type, $uid);
+            }
+        }
+
+        return $variantId;
+    }
+
+    /**
+     * Returns a system unique hash.
+     *
+     * @return string
+     */
+    protected function getSystemHash()
+    {
+        $siteName = isset($GLOBALS['TYPO3_CONF_VARS']['SYS']['sitename']) ? $GLOBALS['TYPO3_CONF_VARS']['SYS']['sitename'] : 'none';
```

Im not sure TYPO3 works with this $GLOBALS['TYPO3_CONF_VARS']['SYS']['sitename'] - but I think I would throw an exception instead